Clean/Utils/BinaryOps.lean

```diff
@@ -0,0 +1,106 @@
+    (∀ x ∈ l, x = 0 ∨ x = 1) → (List.foldl (· &&& ·) 1 l = 0 ∨ List.foldl (· &&& ·) 1 l = 1) := by
+  intro h_all_binary
+  induction l with
+  | nil =>
+    simp only [List.foldl_nil]
+    right; trivial
+  | cons x xs ih =>
+    simp only [List.foldl_cons]
+    have h_x_binary : x = 0 ∨ x = 1 := h_all_binary x (List.Mem.head xs)
+    have h_xs_binary : ∀ y ∈ xs, y = 0 ∨ y = 1 := fun y hy =>
+      h_all_binary y (List.Mem.tail x hy)
+    have h_tail_binary := ih h_xs_binary
+    cases h_x_binary with
+    | inl h_x_zero =>
+      rw [h_x_zero]
+      simp only [HAnd.hAnd, AndOp.and]
+      left
+      suffices h : List.foldl (· &&& ·) (1 &&& 0) xs = 0 by
+        simp only [List.foldl_cons, HAnd.hAnd, AndOp.and] at h ⊢
+        exact h
+      have h_one_zero : (1 : ℕ) &&& 0 = 0 := by
+        simp only [HAnd.hAnd, AndOp.and]
+        rfl
+      rw [h_one_zero]
+      clear h_one_zero h_x_zero h_all_binary h_xs_binary h_tail_binary ih
+      generalize hxs : xs = xs'
+      clear xs hxs
+      induction xs' with
+      | nil => simp [List.foldl_nil]
+      | cons y ys ih =>
+        simp only [List.foldl_cons, HAnd.hAnd, AndOp.and]
+        have h_zero_y : (0 : ℕ).land y = 0 := by
+          unfold Nat.land
+          simp [Nat.bitwise]
+        rw [h_zero_y]
+        exact ih
+    | inr h_x_one =>
+      rw [h_x_one]
+      have h_one_one : (1 : ℕ) &&& 1 = 1 := by
+        simp only [HAnd.hAnd, AndOp.and]
+        rfl
+      rw [h_one_one]
+      exact h_tail_binary
+
+/-- For binary values and binary lists, a &&& foldl orig l = foldl (a &&& orig) l -/
+theorem List.and_foldl_eq_foldl_of_all_binary (a : ℕ) (orig : ℕ) (l : List ℕ)
+    (_ha : a = 0 ∨ a = 1) (hl : ∀ x ∈ l, x = 0 ∨ x = 1) :
+    a &&& List.foldl (· &&& ·) orig l = List.foldl (· &&& ·) (a &&& orig) l := by
+  induction l generalizing orig with
+  | nil =>
+    simp only [List.foldl_nil]
+  | cons hd tl ih =>
+    simp only [List.foldl_cons]
+    have hhd : hd = 0 ∨ hd = 1 := hl hd (List.Mem.head _)
+    have htl : ∀ x ∈ tl, x = 0 ∨ x = 1 := fun x hx => hl x (List.mem_cons_of_mem hd hx)
+    rw [ih (orig &&& hd) htl]
+    congr 1
+    simp only [HAnd.hAnd, AndOp.and]
+    show a &&& (orig &&& hd) = (a &&& orig) &&& hd
+    exact (Nat.land_assoc a orig hd).symm
+
+end ListOperations
+
+section VectorOperations
+
+/-- If all elements of a vector are binary, then all elements of its list are binary -/
+theorem Vector.toList_binary {α : Type*} {n : ℕ} (v : Vector α n) 
+    (isBinary : α → Prop) :
+    (∀ i : Fin n, isBinary (v.get i)) →
+    (∀ x ∈ v.toList, isBinary x) := by
+  intro h_vec x h_mem
+  rw [Vector.mem_toList_iff_get] at h_mem
+  rcases h_mem with ⟨i, hi⟩
+  rw [hi]
+  exact h_vec i
+
+/-- Specialized version for F p where binary means = 0 or = 1 -/
+theorem Vector.toList_binary_field {n : ℕ} (v : Vector (F p) n) :
+    (∀ i : Fin n, v.get i = 0 ∨ v.get i = 1) →
+    (∀ x ∈ v.toList, x = 0 ∨ x = 1) := 
+  Vector.toList_binary v (fun x => x = 0 ∨ x = 1)
```

Comment:
these theorems feel overly specific, and shouldn't be here! use the `mem_iff_getElem`-type theorem directly